punchcut/src/outline.rs

```diff
@@ -0,0 +1,293 @@
+                        break;
+                    }
+                }
+            }
+            if count > 0 && !last_was_close {
+                self.verbs.push(Verb::Close);
+                last_was_close = true;
+            }
+        }
+        true
+    }
+}
+
+impl Shape for Outline {
+    type PathElementsIter<'iter> = Elements<'iter>;
+
+    fn path_elements(&self, _tolerance: f64) -> Self::PathElementsIter<'_> {
+        self.elements()
+    }
+
+    fn to_path(&self, _tolerance: f64) -> BezPath {
+        BezPath::from_vec(self.elements().collect())
+    }
+
+    fn area(&self) -> f64 {
+        segments(self.elements()).map(|seg| seg.signed_area()).sum()
+    }
+
+    fn perimeter(&self, accuracy: f64) -> f64 {
+        segments(self.elements())
+            .map(|seg| seg.arclen(accuracy))
+            .sum()
+    }
+
+    fn winding(&self, pt: Point) -> i32 {
+        segments(self.elements()).map(|seg| seg.winding(pt)).sum()
+    }
+
+    fn bounding_box(&self) -> Rect {
+        let mut bbox: Option<Rect> = None;
+        for seg in segments(self.elements()) {
+            let seg_bb = ParamCurveExtrema::bounding_box(&seg);
+            if let Some(bb) = bbox {
+                bbox = Some(bb.union(seg_bb));
+            } else {
+                bbox = Some(seg_bb)
+            }
+        }
+        bbox.unwrap_or_default()
+    }
+}
+
+/// Iterator over the elements of a path.
+#[derive(Clone)]
+pub struct Elements<'a> {
+    verbs: &'a [Verb],
+    points: &'a [(f32, f32)],
+    verb_pos: usize,
+    point_pos: usize,
+}
+
+impl<'a> Iterator for Elements<'a> {
+    type Item = PathEl;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        fn pt(p: (f32, f32)) -> Point {
+            Point::new(p.0 as f64, p.1 as f64)
+        }
+        let verb = self.verbs.get(self.verb_pos)?;
+        self.verb_pos += 1;
+        Some(match verb {
+            Verb::MoveTo => {
+                let p0 = self.points[self.point_pos];
+                self.point_pos += 1;
+                PathEl::MoveTo(pt(p0))
+            }
+            Verb::LineTo => {
+                let p0 = self.points[self.point_pos];
+                self.point_pos += 1;
+                PathEl::LineTo(pt(p0))
+            }
+            Verb::QuadTo => {
+                let p0 = self.points[self.point_pos];
+                let p1 = self.points[self.point_pos + 1];
+                self.point_pos += 2;
+                PathEl::QuadTo(pt(p0), pt(p1))
+            }
+            Verb::CurveTo => {
+                let p0 = self.points[self.point_pos];
+                let p1 = self.points[self.point_pos + 1];
+                let p2 = self.points[self.point_pos + 2];
+                self.point_pos += 3;
+                PathEl::CurveTo(pt(p0), pt(p1), pt(p2))
+            }
+            Verb::Close => PathEl::ClosePath,
+        })
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+#[repr(u8)]
+enum Verb {
```

Comment:
silly nit, but maybe move this up to be near the `Outline` struct? would improve the experience of reading this file I think